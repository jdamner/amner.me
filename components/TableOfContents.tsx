"use client";

import GithubSlugger from "github-slugger";
import ReactMarkdown from "react-markdown";
import { useEffect, type PropsWithChildren } from "react";

/**
 * Generates an inline table of contents
 * Exclude titles matching this string (new RegExp('^(' + string + ')$', 'i')).
 * If an array is passed the array gets joined with a pipe (new RegExp('^(' + array.join('|') + ')$', 'i')).
 *
 * @param {TOCInlineProps} props
 */
function TableOfContents({ children }: PropsWithChildren) {
  const slugger = new GithubSlugger();

  const toc = children
    .toString()
    .match(/^(#{1,6})\s(.+)$/gm)
    ?.map((heading) => ({
      value: heading.replace(/^(#{1,6})\s/, ""),
      depth: heading.match(/^(#{1,6})\s/)?.[1].length ?? Infinity,
      url: `#${slugger.slug(heading.replace(/^(#{1,6})\s/, ""))}`,
    }));

  useEffect(() => {
    const tocLinks = Array.from(document.querySelectorAll('a[href^="#"]'));
    const headings = Array.from(
      document.querySelectorAll("h1, h2, h3, h4, h5, h6"),
    );

    const observer = new IntersectionObserver(
      (entries) => {
        entries.forEach((entry) => {
          const id = entry.target.getAttribute("id");
          const tocLink = tocLinks.find(
            (link) => link.getAttribute("href") === `#${id}`,
          );
          if (tocLink) {
            if (entry.isIntersecting) {
              tocLink.classList.add("font-bold");
            } else {
              tocLink.classList.remove("font-bold");
            }
          }
        });
      },
      { threshold: 0.5 },
    );

    headings.forEach((heading) => {
      observer.observe(heading);
    });

    return () => {
      headings.forEach((heading) => {
        observer.unobserve(heading);
      });
    };
  });

  if (!toc) return null;

  const minDepth = toc.reduce(
    (min, heading) => Math.min(min, heading.depth),
    6,
  );
  const marginCalc = (depth: number) =>
    `${Math.max(0, depth - (minDepth - 1))}rem`;

  return (
    <div className="py-3">
      <h2 className="text-lg font-black font-sans uppercase my-2">Contents</h2>
      <ul className="text-sm list-disc">
        {toc?.map((heading) =>
          heading.depth < 2 ? (
            <li key={heading.url}>
              <a href={heading.url}>
                <ReactMarkdown className="prose prose-slate dark:prose-invert prose-sm">
                  {heading.value}
                </ReactMarkdown>
              </a>
            </li>
          ) : (
            <li
              key={heading.url}
              style={{ marginLeft: marginCalc(heading.depth) }}
            >
              <a
                href={heading.url}
                onClick={(event) => {
                  event.preventDefault();
                  history.pushState(null, "", heading.url);
                  document
                    .getElementById(heading.url.replace("#", ""))
                    ?.scrollIntoView({ behavior: "smooth" });
                }}
              >
                <ReactMarkdown className="prose prose-slate dark:prose-invert prose-sm">
                  {heading.value}
                </ReactMarkdown>
              </a>
            </li>
          ),
        )}
      </ul>
    </div>
  );
}

export default TableOfContents;
