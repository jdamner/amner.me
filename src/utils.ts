/**
 * Helper functions
 */
import fs from 'fs'
import matter from 'gray-matter'
import { join } from 'path'
import { slug } from 'github-slugger'

import type { ReactNode } from 'react'
import type { MdFile } from './types'

/**
 * Makes an object JSON parseable
 * 
 * @param {any} obj The object to be parsed
 * @returns {any}
 */
export function makeJsonParseable<Type>(obj: Type): Type {
	return JSON.parse(JSON.stringify(obj)) as Type
}

/**
 * Reads a markdown file
 */
export async function readMdFile(filename: string, directory: string = 'content/'): Promise<MdFile> {
	const path = `${directory}/${filename}`
	const fullPath = `${process.cwd()}/${path}`
	const fileContents = fs.readFileSync(fullPath, 'utf8')
	return {
		slug: slug(filename.replace(/\.md$/, '')),
		...matter(fileContents),
	}
}

/**
 * Reads all slugs from a directory
 * 
 * @param {string} directory The directory to read from
 * 
 * @returns {string[]}
 */
export function getFilesInDir(directory: string): string[] {
	const fullPath = join(process.cwd(), directory)
	return fs.readdirSync(fullPath)
}

/**
 * Converts all headings in a markdown file to have IDs 
 * Converts 
 * ## Heading 1 
 * to become
 * ## Heading 1{#heading-1}
 * 
 * @param {string} markdown The markdown to convert
 * 
 * @returns {string}
 */
export function convertHeadingsToAnchorLinks(markdown: string): string {
	return markdown.replace(/(#{1,6})\s(.+)\n/gm, (match, p1, p2) => {
		return `${p1} ${p2}{#${slug(p2)}}\n`
	})
}


/**
 * Converts a react node to a slug
 * 
 * Usually the value is a string, but sometimes it contains
 * a `<code>` element, which is a react node. This function
 * converts the value to a string.
 */
export function makeSlug(value: ReactNode | ReactNode[]): string {
	if (! Array.isArray(value)) return slug(value?.toString() ?? '');
	return slug(
		value.map((node) => {
			if (typeof node === 'string' ||
				typeof node === 'number' ||
				typeof node === 'boolean'
			) return node;

			if (typeof node === 'object' && node !== null && 'props' in node) {
				return makeSlug(node.props.children)
			}

			return ''
		}).join('')
	)
}