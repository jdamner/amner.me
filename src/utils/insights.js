import { track, parameters, init } from 'insights-js';

const pageView = () => {
    event('page-view');
}

const event = (eventName, eventProperties = {}) => {
    init('HbSIygPWVvMzXFRP');
    track({
        id: eventName,
        parameters: {
            ...parameters,
            ...eventProperties,
        },
    });
}

export { pageView, event };