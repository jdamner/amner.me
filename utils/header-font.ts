import { Bebas_Neue as Font } from "next/font/google";
export const font = Font({
  weight: "400",
  subsets: ["latin", "latin-ext"],
});
