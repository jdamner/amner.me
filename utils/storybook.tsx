import { faker } from "@faker-js/faker";
import { MdFile, PostLinkType } from "../types";
import matter from "gray-matter";
import React from "react";

type MdAttributes = Record<string, string>;

faker.seed(123);

export const generateMdFile = <T extends MdAttributes>(
  attributes: T,
  content?: string,
): Omit<MdFile<string>, "data"> & { data: T } => {
  return matter(generateMD(attributes, content)) as Omit<
    MdFile<string>,
    "data"
  > & { data: T };
};

export const generateMdContent = (): string => {
  return (
    `# ${faker.lorem.words(3)}\n` +
    Array.from({ length: 3 }, () => {
      return (
        `## ${faker.lorem.words(2)}\n` +
        faker.lorem.paragraph() +
        "\n" +
        Array.from({ length: 3 }, () => {
          return (
            `### ${faker.lorem.words(2)}\n` + `${faker.lorem.paragraph()}\n`
          );
        }).join("")
      );
    }).join("")
  );
};

export const generatePostLink = (): PostLinkType => {
  return {
    title: faker.lorem.words(3),
    thumbnail: faker.image.url(),
    slug: faker.lorem.slug(),
  };
};

export const generatePostLinks = (count: number): PostLinkType[] => {
  return Array.from({ length: count }, () => generatePostLink());
};

export const generateMD = (
  attributes: MdAttributes,
  content?: string,
): string => {
  const frontMatter = generateMDFrontMatter(attributes);
  return `${frontMatter}\n${content ?? generateMdContent()}`;
};

export const generateMDFrontMatter = (attributes: MdAttributes): string => {
  return (
    "---\n" +
    Object.keys(attributes)
      .map((key) => `${key}: ${attributes[key]}`)
      .join("\n") +
    "\n---"
  );
};

export const generatePlaceholderContent = (
  size: number = 5,
): React.ReactNode => {
  return (
    <>
      {Array.from({ length: size }).map((_, i) => (
        <p key={i}>{faker.lorem.paragraph(i + 1)}</p>
      ))}
    </>
  );
};
