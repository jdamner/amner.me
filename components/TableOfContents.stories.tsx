import type { Meta, StoryObj } from "@storybook/react";

import TableOfContents from "./TableOfContents";
import { generateMdContent } from "../utils/storybook";

type Story = StoryObj<typeof TableOfContents>;

export default {
  title: "Components/Table of Contents",
  component: TableOfContents,
} as Meta;

export const Default: Story = {
  args: {
    children: generateMdContent(),
  },
};
