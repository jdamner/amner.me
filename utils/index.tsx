/* External Deps */
import { readFileSync, readdirSync } from "fs";
import { resolve } from "path";
import matter from "gray-matter";
import { slug } from "github-slugger";

/* Components */
import { Markdown } from "@/components/Markdown";

/* Constants */
const postsDirectory: string = resolve("content/blog/");
const servicesDirectory: string = resolve("content/services");
const cvDirectory: string = resolve("content/cv");
const projectsDirectory: string = resolve(cvDirectory, "projects");
const referencesDirectory: string = resolve(cvDirectory, "references");
const employmentDirectory: string = resolve(cvDirectory, "employment");
const educationDirectory: string = resolve(cvDirectory, "education");

type MatterData = ReturnType<typeof matter>["data"];
type AttributeValidator<T extends MatterData> = (data: MatterData) => data is T;

const defaultAttributeValidator = <T extends MatterData>(
  data: MatterData,
): data is T => {
  return typeof data === "object" && data !== null;
};

/**
 * Gets all files in a directory
 */
const getAllFrom = <T extends MatterData>(
  directory: string,
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) =>
  readdirSync(directory).map((filename) =>
    readMdFile<T>(filename, directory, attributeValidator),
  );

/**
 * Reads a markdown file
 */
const readMdFile = <T extends MatterData>(
  filename: string,
  directory: string = "content/",
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) =>
  parseMarkdown<T>(
    readFileSync(`${directory}/${filename}`, "utf8"),
    filename,
    attributeValidator,
  );

/**
 * Parses markdown file contents
 */
export const parseMarkdown = <T extends MatterData>(
  fileContents: string,
  filename: string,
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) => {
  const { content, data } = matter(fileContents);

  if (!attributeValidator(data)) {
    console.log(data);
    throw new Error(`Invalid attributes in ${filename}` + JSON.stringify(data));
  }

  return {
    data,
    slug: slug(filename.replace(/\.mdx$/, "")),
    content: <Markdown>{content}</Markdown>,
  };
};

/**
 * Gets all Posts
 */
const getAllPosts = () =>
  getAllFrom<WithTitle & WithThumbnail & WithDate>(
    postsDirectory,
    (data): data is WithTitle & WithThumbnail & WithDate =>
      hasTitle(data) && hasThumbnail(data) && hasDate(data),
  );

export const useAllServices = () =>
  getAllFrom<WithTitle>(servicesDirectory, hasTitle);

export const getAllProjects = () =>
  getAllFrom<WithTitle & WithLink>(
    projectsDirectory,
    (data): data is WithTitle & WithLink => hasTitle(data) && hasLink(data),
  );

export const useAllReferences = () =>
  getAllFrom<WithTitle & WithWebsite & WithPhone & WithEmail>(
    referencesDirectory,
    (data): data is WithTitle & WithWebsite & WithPhone & WithEmail =>
      hasTitle(data) && hasWebsite(data) && hasPhone(data) && hasEmail(data),
  );

export const useAllEmployment = () =>
  getAllFrom<
    WithTitle & WithSubtitle & WithDateString & WithRole & WithWebsite
  >(
    employmentDirectory,
    (
      data,
    ): data is WithTitle &
      WithSubtitle &
      WithDateString &
      WithRole &
      WithWebsite =>
      hasTitle(data) &&
      hasSubtitle(data) &&
      hasDateString(data) &&
      hasRole(data) &&
      hasWebsite(data),
  );

export const useAllEducation = () =>
  getAllFrom<WithTitle & WithDateString>(
    educationDirectory,
    (data): data is WithTitle & WithDateString =>
      hasTitle(data) && hasDateString(data),
  );

export const getPost = (slug: string) =>
  getAllPosts().find((post) => post.slug === slug);

export const getAllPostLinks = () =>
  getAllPosts()
    .map((post) => ({
      title: post.data.title,
      thumbnail: post.data.thumbnail,
      slug: post.slug,
      date: post.data.date,
    }))
    .sort((a, b) => (a.date > b.date ? -1 : 1));

type WithTitle = { title: string };
type WithSubtitle = { subtitle: string };
type WithLink = { link: string };
type WithThumbnail = { thumbnail: string };
type WithDate = { date: Date };
type WithDateString = { date: string };
type WithPhone = { phone: string };
type WithWebsite = { website: string };
type WithEmail = { email: string };
type WithRole = { role: string };

const hasTitle = (data: MatterData): data is WithTitle =>
  defaultAttributeValidator(data) &&
  "title" in data &&
  typeof data.title === "string";

const hasSubtitle = (data: MatterData): data is WithSubtitle =>
  "subtitle" in data && typeof data.subtitle === "string";

const hasLink = (data: MatterData): data is WithLink =>
  "link" in data && typeof data.link === "string";

const hasThumbnail = (data: MatterData): data is WithThumbnail =>
  "thumbnail" in data && typeof data.thumbnail === "string";

const hasDate = (data: MatterData): data is WithDate =>
  "date" in data && data.date instanceof Date;

const hasDateString = (data: MatterData): data is WithDateString =>
  "date" in data && typeof data.date === "string";

const hasPhone = (data: MatterData): data is WithPhone =>
  "phone" in data && typeof data.phone === "string";

const hasWebsite = (data: MatterData): data is WithWebsite =>
  "website" in data && typeof data.website === "string";

const hasEmail = (data: MatterData): data is WithEmail =>
  "email" in data && typeof data.email === "string";

export const hasRole = (data: MatterData): data is WithRole =>
  "role" in data && typeof data.role === "string";
