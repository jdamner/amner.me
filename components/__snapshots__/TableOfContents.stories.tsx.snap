// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Table of Contents Default smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <div class="py-3">
    <h2 class="text-lg font-black font-sans uppercase my-2">
      Contents
    </h2>
    <ul class="text-sm list-disc">
      <li>
        <a href="#subito-circumvenio-calamitas">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              subito circumvenio calamitas
            </p>
          </div>
        </a>
      </li>
      <li style="margin-left: 2rem;">
        <a href="#ipsa-super">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              ipsa super
            </p>
          </div>
        </a>
      </li>
      <li style="margin-left: 3rem;">
        <a href="#calco-claudeo">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              calco claudeo
            </p>
          </div>
        </a>
      </li>
      <li style="margin-left: 3rem;">
        <a href="#stabilis-utique">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              stabilis utique
            </p>
          </div>
        </a>
      </li>
      <li style="margin-left: 3rem;">
        <a href="#stultus-iste">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              stultus iste
            </p>
          </div>
        </a>
      </li>
      <li style="margin-left: 2rem;">
        <a href="#solvo-tyrannus">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              solvo tyrannus
            </p>
          </div>
        </a>
      </li>
      <li style="margin-left: 3rem;">
        <a href="#arto-paulatim">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              arto paulatim
            </p>
          </div>
        </a>
      </li>
      <li style="margin-left: 3rem;">
        <a href="#teneo-maxime">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              teneo maxime
            </p>
          </div>
        </a>
      </li>
      <li style="margin-left: 3rem;">
        <a href="#titulus-absens">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              titulus absens
            </p>
          </div>
        </a>
      </li>
      <li style="margin-left: 2rem;">
        <a href="#cupiditas-denique">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              cupiditas denique
            </p>
          </div>
        </a>
      </li>
      <li style="margin-left: 3rem;">
        <a href="#vomer-fugit">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              vomer fugit
            </p>
          </div>
        </a>
      </li>
      <li style="margin-left: 3rem;">
        <a href="#vulgo-capillus">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              vulgo capillus
            </p>
          </div>
        </a>
      </li>
      <li style="margin-left: 3rem;">
        <a href="#adipisci-sum">
          <div class="prose prose-slate dark:prose-invert prose-sm">
            <p>
              adipisci sum
            </p>
          </div>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
